planner/core/exhaust_physical_plans.go

```diff
@@ -2351,23 +2342,23 @@ func (la *LogicalAggregation) getHashAggs(prop *property.PhysicalProperty) []Phy
 		taskTypes = append(taskTypes, property.CopTiFlashLocalReadTaskType)
 	}
 	canPushDownToMPP := la.ctx.GetSessionVars().AllowMPPExecution && !collate.NewCollationEnabled() && la.checkCanPushDownToMPP()
-	if canPushDownToMPP {
-		taskTypes = append(taskTypes, property.MppTaskType)
-	}
 	if la.HasDistinct() {
 		// TODO: remove after the cost estimation of distinct pushdown is implemented.
-		if !la.ctx.GetSessionVars().AllowDistinctAggPushDown && !canPushDownToMPP {
+		if !la.ctx.GetSessionVars().AllowDistinctAggPushDown {
 			taskTypes = []property.TaskType{property.RootTaskType}
 		}
 	} else if !la.aggHints.preferAggToCop {
 		taskTypes = append(taskTypes, property.RootTaskType)
 	}
+	if !la.canPushToCop() {
+		taskTypes = []property.TaskType{property.RootTaskType, property.CopTiFlashLocalReadTaskType}
+	}
```

Comment:
How about moving this logic to line 2340 since it's used to initialize `taskTypes`?

Reply:
We can't do it for now. If we consider the root task in line 2340, we might append root task twice. If not, it's possible that there is no root task if allow distinct push down or agg_cop. That's not perfect but let's leave it be for now,